src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.monoid_algebra.basic
+import data.multiset.lattice
+/-!  #  Degree of an `add_monoid_algebra`
+
+Let `R` be a semiring and let `A` be a `semilattice_sup`.
+
+For an element `f : add_monoid_algebra R A`, this file defines
+* `add_monoid_algebra.degree`: the degree taking values in `with_bot A`,
+* `add_monoid_algebra.trailing_degree`: the trailing degree taking values in `with_top A`.
+If the grading type `A` is a linearly ordered additive monoid, then this notions of degree
+coincide with the standard one:
+* the degree is the maximum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊥`, if `f = 0`;
+* the trailing degree is the minimum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊤`, if `f = 0`.
+
+Currently, the only results are
+* `degree_mul_le` -- the degree of a product is at most the sum of the degrees,
+*  `le_trailing_degree_mul` -- the trailing degree of a product is at least the sum of the
+  trailing degrees.
+-/
+
+namespace add_monoid_algebra
+open_locale classical
+
+variables {R A : Type*} [semiring R]
+
+section degree
+variables [semilattice_sup A]
+
+/--  Let `R` be a semiring, let `A` be a `semilattice_sup`, and let `f : R[A]` be an
+an element of `add_monoid_algebra R A`.  The degree of `f` takes values in `with_bot A` and
+it is the supremum of the support of `f` or `⊥`, depending on whether `f` is non-zero or not.
+
+If `A` has a linear order, then this notion coincides with the usual one, using the maximum of
+the exponents. -/
+def degree (f : add_monoid_algebra R A) : with_bot A :=
+f.support.sup coe
```

Comment:
You proved everything for `id` earlier; is `degree_mul_le` true for any other functions? That is, can you first prove
```
lemma sup_support_mul_le [order_bot B] (f g : add_monoid_algebra R A) (h : A \to B):
  (f * g).support.sup h ≤ f.support.sup h + g.support.sup h :=
```
with some conditions on `h`?

Reply:
(application: the maximum degree of a multivariate polynomial)

Reply:
Very likely, if `f` is "monotone" (whatever the correct property is for `semilattice_sup`), the same would hold.  I'll try it.

Reply:
@eric-wieser, I just pushed a version of the definitions that uses a "generic" weight function on the grading that is assumed to be essentially "convex" (or sublinear or the opposite, I never pay enough attention to which way the inequalities go).

Is this what you had in mind?

As a sanity check, I left the old definitions, in terms of `sup coe` and showed that they are defeq to the newer ones, in their more restricted scope.

I wonder if I should give this "weighted degree" a different name as still call `degree` its specialized version.  Any suggestions/thoughts are welcome!